### DISCUSSION TITLE
RavenDB for Uniyt3D

### DISCUSSION THREAD

**Discussion by LabLENI** [ravendb/ravendb#19677] [Q&A]
Hello,

My name is José, and I belong to a research institute at the Polytechnic University of Valencia.
The features offered by RavenDB are very interesting for the type of projects/studies we conduct.
Our idea is to use it with .NET, Python, and the prototyping tool Unity3D.
With .NET in Visual Studio 2019, everything works smoothly, but we are having issues using it with Unity3D.
I have tested it with Unity3D version 6.
I install the RavenDB.Client 6.2.1 client via NuGet For Unity, but when I open the session:
IDocumentSession session = store.OpenSession();
It throws an exception related to the certificate.

With the client version 5.2, I can initialize the store and query documents, and it correctly lists the content of a document. However, it crashes when I execute session.Store(xxx) to perform insertions/updates.
The error is: The type initializer for 'Sparrow.Utils.DecimalHelper' threw an exception.

This same code works perfectly in a console application with .NET.

Is there any kind of client, even with reduced functionality, that is compatible with .net in Unity3D?

Thank you and best regards,

José

**Comment by ppekrol** [comment]
Hi,

could I ask you to setup a simple sample app that reproduces the problem? We would be happy to take a peek and fix the issue and direct repro will speed up the process.

**Comment by LabLENI** [comment]
Okay, thank you very much.
You only need Unity3D 6 -> https://unity.com/releases/editor/whats-new/6000.0.28#installs
NuGet for Unity -> https://github.com/GlitchEnzo/NuGetForUnity/releases/tag/v4.1.1
To install RavenDB.Client
[RavenDBManager.zip](https://github.com/user-attachments/files/17936393/RavenDBManager.zip)

And this simple script on any Unity object (attached)
Then complete the "databaseUrl" and "databaseName", and if necessary, the path to the certificate "certPath".

**Reply by ppekrol** [reply]
hi @LabLENI 

I checked your solution (started fresh project, using Unity 6 6000.0.28f1) and it works without issues. Persisted everything correctly:
![image](https://github.com/user-attachments/assets/52b1397d-d4b0-4ebc-855d-6570a0dd7418)

Do I need to make any changes in the code? Change some runtime maybe?

The piece of code that fails for you is:
```csharp
                string flagsFieldName;
                var decimalType = typeof(decimal);
                if (decimalType.GetField("_flags", BindingFlags.Instance | BindingFlags.NonPublic) != null)
                    flagsFieldName = "_flags";
                else if (decimalType.GetField("flags", BindingFlags.Instance | BindingFlags.NonPublic) != null)
                    flagsFieldName = "flags";
                else
                    throw new InvalidOperationException("Could not determine name of the 'flags' field in decimal type");
```

For me it returns `flags` .

**Comment by LabLENI** [comment]
I pointed to the test database at http://live-test.ravendb.net/ and with a clean database (test4), the first insertion worked. But on the second run, it failed.

My .NET configuration in Unity 6 is:

![image](https://github.com/user-attachments/assets/577efff9-b6a8-4135-a281-201c025398b3)


And the versions I have tried are the latest available:

RavenDB.Client 6.2.1

Here I encounter an issue: -> List<Persona> personas = session.Query<Persona>().ToList();

Error: Index was outside the bounds of the array.

And with RavenDB.Client 5.2.100  here: -> session.Store(persona); // Save the new document in the session

Error: The type initializer for 'Sparrow.Utils.DecimalHelper' threw an exception.
UnityEngine.Debug:LogError (object)

RavenDBManager:Start () (at Assets/_scripts/RavenDBManager.cs:101)


What version of the RavenDB.Client package have you used?

**Reply by ppekrol** [reply]
@LabLENI 

I successfully ran your code with 5.4.204 Client API. Can you try that?

I'm pretty sure that 5.2.100 had a bug with that DecimalHelper and it was fixed afterwards.

On 6.2.1 I'm getting IndexOutOfRangeException. I will be looking into that, it has to do sth with Mono IMO.

**Comment by LabLENI** [comment]
With Unity 6 and the client version you mentioned, it works!  
I will run more tests and I also need to try using a web client from Unity with the REST API. However, in that case, I think the issues will arise when passing the certificate. For those matters, I’ll create another post if necessary. 

Thank you very much!
